pmtiles/bucket.go

```diff
@@ -2,64 +2,117 @@ package pmtiles
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"gocloud.dev/blob"
 	"io"
 	"net/http"
 	"net/url"
 	"os"
 	"path"
 	"path/filepath"
 	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"gocloud.dev/blob"
 )
 
 // Bucket is an abstration over a gocloud or plain HTTP bucket.
 type Bucket interface {
 	Close() error
 	NewRangeReader(ctx context.Context, key string, offset int64, length int64) (io.ReadCloser, error)
+	NewRangeReaderEtag(ctx context.Context, key string, offset int64, length int64, etag string) (io.ReadCloser, string, error)
+}
+
+type RefreshRequiredError struct {
+	StatusCode int
+}
+
+func (m *RefreshRequiredError) Error() string {
+	return fmt.Sprintf("HTTP error indicates file has changed: %d", m.StatusCode)
 }
 
 type HTTPBucket struct {
 	baseURL string
 }
 
-func (b HTTPBucket) NewRangeReader(_ context.Context, key string, offset, length int64) (io.ReadCloser, error) {
+func (b HTTPBucket) NewRangeReader(ctx context.Context, key string, offset, length int64) (io.ReadCloser, error) {
+	body, _, err := b.NewRangeReaderEtag(ctx, key, offset, length, "")
+	return body, err
+}
+
+func (b HTTPBucket) NewRangeReaderEtag(_ context.Context, key string, offset, length int64, etag string) (io.ReadCloser, string, error) {
 	reqURL := b.baseURL + "/" + key
 
 	req, err := http.NewRequest("GET", reqURL, nil)
```

Comment:
Is there a reason why this shouldn't use `NewRequestWithContext` and pass through the context variable?

Reply:
It should probably use that instead.